src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluderTest.kt

```diff
@@ -0,0 +1,151 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service
+
+import com.nhaarman.mockitokotlin2.firstValue
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.same
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.verifyZeroInteractions
+import com.nhaarman.mockitokotlin2.whenever
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.mockito.ArgumentCaptor
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events.ReferralEventPublisher
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events.ReferralEventType
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Referral
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ActionPlanAppointmentRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ReferralRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.ActionPlanFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.EndOfServiceReportFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.ReferralFactory
+import java.time.OffsetDateTime
+
+internal class ReferralConcluderTest {
+
+  private val referralRepository: ReferralRepository = mock()
+  private val actionPlanAppointmentRepository: ActionPlanAppointmentRepository = mock()
+  private val referralEventPublisher: ReferralEventPublisher = mock()
+
+  private val referralFactory = ReferralFactory()
+  private val actionPlanFactory = ActionPlanFactory()
+  private val endOfServiceReportFactory = EndOfServiceReportFactory()
+
+  private val referralConcluder = ReferralConcluder(
+    referralRepository, actionPlanAppointmentRepository, referralEventPublisher
+  )
+
+  @Test
+  fun `concludes referral as cancelled when ending a referral with no action plan`() {
+
+    val timeAtStart = OffsetDateTime.now()
+    val referralWithNoActionPlan = referralFactory.createSent()
+
+    referralConcluder.concludeIfEligible(referralWithNoActionPlan)
+
+    verifySaveWithConcludedAtSet(referralWithNoActionPlan, timeAtStart)
+    verifyEventPublished(referralWithNoActionPlan, ReferralEventType.CANCELLED)
+  }
+
+  @Test
+  fun `concludes referral as cancelled when ending a referral with no appointments attended`() {
+
+    val timeAtStart = OffsetDateTime.now()
+    val referralWithActionPlanAndNoAttendedAppointments = referralFactory.createSent()
+    whenever(actionPlanAppointmentRepository.countByActionPlanIdAndAttendedIsNotNull(referralWithActionPlanAndNoAttendedAppointments.id)).thenReturn(0)
```

Comment:
does this test do what it says on the tin? won't the action plan still be null in the referral?

Reply:
You're right. I've missed the creation of an action plan step. Will correct test.